File: cpp/leetcode/251.cpp
class Vector2D {
public:
    Vector2D(vector<vector<int>>& vec2d) {
        start = vec2d.begin();
        end = vec2d.end();
        i = 0;
    }

    int next() {
        if(hasNext()) return (*start)[i++];
        return -1;
    }

    bool hasNext() {
        while(start != end && i == (*start).size()){
            start++;
            i = 0;
        }
        return start != end;
    }
private:
    vector<vector<int>>::iterator start, end;
    int i;
};

File: python/leetcode/251.py
class Vector2D:
    def __init__(self, vec2d):
        self.col = 0
        self.row = 0
        self.vec = vec2d

    def next(self):
        res = self.vec[self.row][self.col]
        self.col += 1
        return res

    def hasNext(self):
        while self.row < len(self.vec):
            if self.col < len(self.vec[self.row]): return True
            self.col = 0
            self.row += 1
        return False